File: src/trapezoidal_map.cpp
#include "trapezoidal_map.h"

#include <algorithm>
#include <cassert>
#include <stdexcept>

#include "bounding_box.h"

/* Linear congruential random number generator.
 * Edges in the triangulation are randomly shuffled
 * before being added to the trapezoid map.
 * Want the shuffling to be identical across different operating systems
 * and the same regardless of previous random number use.
 * Would prefer to use a STL or Boost random number generator,
 * but support is not consistent across different operating systems
 * so implementing own here.
 *
 * This is not particularly random, but is perfectly adequate for the use here.
 * Coefficients taken from Numerical Recipes in C. */
class RandomNumberGenerator {
 public:
  RandomNumberGenerator(unsigned long seed)
      : _m(21870), _a(1291), _c(4621), _seed(seed % _m) {}

  // Return random integer in the range 0 to max_value-1.
  unsigned long operator()(unsigned long max_value) {
    _seed = (_seed * _a + _c) % _m;
    return (_seed * max_value) / _m;
  }

 private:
  const unsigned long _m, _a, _c;
  unsigned long _seed;
};

TrapezoidalMap::TrapezoidalMap(const std::vector<Point>& points)
    : npoints(points.size()), _points(points), _root(nullptr) {
  // Set up points array, which contains all of the points in the
  // triangulation plus the 4 corners of the enclosing rectangle.
  BoundingBox bbox;
  for (auto& xy : _points) {
    // Avoid problems with -0. values different from 0.
    if (xy.x == -0.) xy.x = 0.;
    if (xy.y == -0.) xy.y = 0.;
    bbox.add(xy);
  }

  // Last 4 points are corner points of enclosing rectangle.  Enclosing
  // rectangle made slightly larger in case corner points are already in the
  // triangulation.
  if (bbox.empty) {
    bbox.add(Point(0., 0.));
    bbox.add(Point(1., 1.));
  } else {
    const double small = 0.1;  // Any value > 0.
    bbox.expand((bbox.upper - bbox.lower) * small);
  }
  _points.push_back(Point(bbox.lower));                  // SW point.
  _points.push_back(Point(bbox.upper.x, bbox.lower.y));  // SE point.
  _points.push_back(Point(bbox.lower.x, bbox.upper.y));  // NW point.
  _points.push_back(Point(bbox.upper));                  // NE point.
}

TrapezoidalMap::~TrapezoidalMap() {
  _edges.clear();

  delete _root;
  _root = nullptr;
}

bool TrapezoidalMap::add_edge(const Edge& edge) {
  std::vector<Trapezoid*> trapezoids;
  if (!find_trapezoids_intersecting_edge(edge, trapezoids)) return false;
  assert(!trapezoids.empty() && "No trapezoids intersect edge");

  const Point* p = edge.left;
  const Point* q = edge.right;
  Trapezoid* left_old = nullptr;    // old trapezoid to the left.
  Trapezoid* left_below = nullptr;  // below trapezoid to the left.
  Trapezoid* left_above = nullptr;  // above trapezoid to the left.

  // Iterate through trapezoids intersecting edge from left to right.
  // Replace each old trapezoid with 2+ new trapezoids, and replace its
  // corresponding nodes in the search graph with new nodes.
  std::size_t ntraps = trapezoids.size();
  for (std::size_t i = 0; i < ntraps; ++i) {
    Trapezoid* old = trapezoids[i];  // old trapezoid to replace.
    bool start_trap = (i == 0);
    bool end_trap = (i == ntraps - 1);
    bool have_left = (start_trap && edge.left != old->left);
    bool have_right = (end_trap && edge.right != old->right);

    // Old trapezoid is replaced by up to 4 new trapezoids: left is to the
    // left of the start point p, below/above are below/above the edge
    // inserted, and right is to the right of the end point q.
    Trapezoid* left = nullptr;
    Trapezoid* below = nullptr;
    Trapezoid* above = nullptr;
    Trapezoid* right = nullptr;

    // There are 4 different cases here depending on whether the old
    // trapezoid in question is the start and/or end trapezoid of those
    // that intersect the edge inserted.  There is some code duplication
    // here but it is much easier to understand this way rather than
    // interleave the 4 different cases with many more if-statements.
    if (start_trap && end_trap) {
      // Edge intersects a single trapezoid.
      if (have_left) left = new Trapezoid(old->left, p, old->below, old->above);
      below = new Trapezoid(p, q, old->below, edge);
      above = new Trapezoid(p, q, edge, old->above);
      if (have_right)
        right = new Trapezoid(q, old->right, old->below, old->above);

      // Set pairs of trapezoid neighbours.
      if (have_left) {
        left->set_lower_left(old->lower_left);
        left->set_upper_left(old->upper_left);
        left->set_lower_right(below);
        left->set_upper_right(above);
      } else {
        below->set_lower_left(old->lower_left);
        above->set_upper_left(old->upper_left);
      }

      if (have_right) {
        right->set_lower_right(old->lower_right);
        right->set_upper_right(old->upper_right);
        below->set_lower_right(right);
        above->set_upper_right(right);
      } else {
        below->set_lower_right(old->lower_right);
        above->set_upper_right(old->upper_right);
      }
    } else if (start_trap) {
      // Old trapezoid is the first of 2+ trapezoids that the edge
      // intersects.
      if (have_left) left = new Trapezoid(old->left, p, old->below, old->above);
      below = new Trapezoid(p, old->right, old->below, edge);
      above = new Trapezoid(p, old->right, edge, old->above);

      // Set pairs of trapezoid neighbours.
      if (have_left) {
        left->set_lower_left(old->lower_left);
        left->set_upper_left(old->upper_left);
        left->set_lower_right(below);
        left->set_upper_right(above);
      } else {
        below->set_lower_left(old->lower_left);
        above->set_upper_left(old->upper_left);
      }

      below->set_lower_right(old->lower_right);
      above->set_upper_right(old->upper_right);
    } else if (end_trap) {
      // Old trapezoid is the last of 2+ trapezoids that the edge
      // intersects.
      if (left_below->below == old->below) {
        below = left_below;
        below->right = q;
      } else
        below = new Trapezoid(old->left, q, old->below, edge);

      if (left_above->above == old->above) {
        above = left_above;
        above->right = q;
      } else
        above = new Trapezoid(old->left, q, edge, old->above);

      if (have_right)
        right = new Trapezoid(q, old->right, old->below, old->above);

      // Set pairs of trapezoid neighbours.
      if (have_right) {
        right->set_lower_right(old->lower_right);
        right->set_upper_right(old->upper_right);
        below->set_lower_right(right);
        above->set_upper_right(right);
      } else {
        below->set_lower_right(old->lower_right);
        above->set_upper_right(old->upper_right);
      }

      // Connect to new trapezoids replacing prevOld.
      if (below != left_below) {
        below->set_upper_left(left_below);
        if (old->lower_left == left_old)
          below->set_lower_left(left_below);
        else
          below->set_lower_left(old->lower_left);
      }

      if (above != left_above) {
        above->set_lower_left(left_above);
        if (old->upper_left == left_old)
          above->set_upper_left(left_above);
        else
          above->set_upper_left(old->upper_left);
      }
    } else {  // Middle trapezoid.
      // Old trapezoid is neither the first nor last of the 3+ trapezoids
      // that the edge intersects.
      if (left_below->below == old->below) {
        below = left_below;
        below->right = old->right;
      } else
        below = new Trapezoid(old->left, old->right, old->below, edge);

      if (left_above->above == old->above) {
        above = left_above;
        above->right = old->right;
      } else
        above = new Trapezoid(old->left, old->right, edge, old->above);

      // Connect to new trapezoids replacing prevOld.
      if (below != left_below) {  // below is new.
        below->set_upper_left(left_below);
        if (old->lower_left == left_old)
          below->set_lower_left(left_below);
        else
          below->set_lower_left(old->lower_left);
      }

      if (above != left_above) {  // above is new.
        above->set_lower_left(left_above);
        if (old->upper_left == left_old)
          above->set_upper_left(left_above);
        else
          above->set_upper_left(old->upper_left);
      }

      below->set_lower_right(old->lower_right);
      above->set_upper_right(old->upper_right);
    }

    // Create new nodes to add to search graph.  Below and above trapezoids
    // may already have owning trapezoid nodes, in which case reuse them.
    Node* new_top_node = new Node(
        &edge, below == left_below ? below->trapezoid_node : new Node(below),
        above == left_above ? above->trapezoid_node : new Node(above));
    if (have_right) new_top_node = new Node(q, new_top_node, new Node(right));
    if (have_left) new_top_node = new Node(p, new Node(left), new_top_node);

    // Insert new_top_node in correct position or positions in search graph.
    Node* old_node = old->trapezoid_node;
    if (old_node == _root)
      _root = new_top_node;
    else
      old_node->replace_with(new_top_node);

    // old_node has been removed from all of its parents and is no longer
    // needed.
    assert(old_node->has_no_parents() && "Node should have no parents");
    delete old_node;

    // Clearing up.
    if (!end_trap) {
      // Prepare for next loop.
      left_old = old;
      left_above = above;
      left_below = below;
    }
  }

  return true;
}

bool TrapezoidalMap::find_trapezoids_intersecting_edge(
    const Edge& edge, std::vector<Trapezoid*>& trapezoids) {
  // This is the FollowSegment algorithm of de Berg et al, with some extra
  // checks to deal with simple collinear (i.e. invalid) triangles.
  trapezoids.clear();
  Trapezoid* trapezoid = _root->search(edge);
  if (trapezoid == nullptr) {
    assert(trapezoid != nullptr && "search(edge) returns null trapezoid");
    return false;
  }

  trapezoids.push_back(trapezoid);
  while (edge.right->is_right_of(*trapezoid->right)) {
    int orient = edge.get_point_orientation(*trapezoid->right);
    if (orient == 0) {
      assert(0 && "Unable to deal with point on edge");
      return false;
    }

    if (orient == -1)
      trapezoid = trapezoid->lower_right;
    else if (orient == +1)
      trapezoid = trapezoid->upper_right;

    if (trapezoid == nullptr) {
      assert(0 && "Expected trapezoid neighbor");
      return false;
    }
    trapezoids.push_back(trapezoid);
  }

  return true;
}

void TrapezoidalMap::build() {
  // Set up edges array.
  // First the bottom and top edges of the enclosing rectangle.
  _edges.push_back(Edge(&_points[npoints], &_points[npoints + 1]));
  _edges.push_back(Edge(&_points[npoints + 2], &_points[npoints + 3]));

  for (std::size_t index = 0; index < npoints; ++index) {
    Point* start = &_points[index];
    Point* end = &_points[(index + 1) % npoints];
    if (end->is_right_of(*start))
      _edges.push_back(Edge(start, end));
    else
      _edges.push_back(Edge(end, start));
  }

  // Initial trapezoid is enclosing rectangle.
  _root = new Node(new Trapezoid(&_points[npoints], &_points[npoints + 1],
                                 _edges[0], _edges[1]));
  _root->assert_valid();

  // Randomly shuffle all edges other than first 2.
  RandomNumberGenerator rng(1234);
  std::random_shuffle(_edges.begin() + 2, _edges.end(), rng);

  // Add edges, one at a time, to graph.
  std::size_t nedges = _edges.size();
  for (std::size_t index = 2; index < nedges; ++index) {
    if (!add_edge(_edges[index]))
      throw std::runtime_error("Triangulation is invalid");
    _root->assert_valid();
  }
}

File: seidel/trapezoidal_map.py
import random
from typing import (List,
                    Sequence)

from .bounding_box import BoundingBox
from .edge import Edge
from .leaf import Leaf
from .node import Node
from .point import Point
from .trapezoid import Trapezoid
from .x_node import XNode
from .y_node import YNode


def build_graph(points: Sequence[Point], shuffle: bool) -> Node:
    bounding_box = BoundingBox(True, Point(0, 0), Point(0, 0))
    points = list(points)
    points_count = len(points)
    for point in points:
        if not point.x:
            point.x = 0
        if not point.y:
            point.y = 0
        bounding_box.add(point)
    # Last 4 points are corner points of enclosing rectangle.  Enclosing
    # rectangle made slightly larger in case corner points are already in the
    # triangulation.
    if bounding_box.empty:
        bounding_box.add(Point(0., 0.))
        bounding_box.add(Point(1., 1.))
    else:
        small = 0.1  # Any value > 0.
        bounding_box.expand((bounding_box.upper - bounding_box.lower) * small)
    # SW point.
    points.append(bounding_box.lower)
    # SE point.
    points.append(Point(bounding_box.upper.x, bounding_box.lower.y))
    # NW point.
    points.append(Point(bounding_box.lower.x, bounding_box.upper.y))
    # NE point.
    points.append(bounding_box.upper)

    edges = [Edge(points[points_count], points[points_count + 1]),
             Edge(points[points_count + 2], points[points_count + 3])]
    for index in range(points_count):
        start, end = points[index], points[(index + 1) % points_count]
        edges.append(Edge(start, end)
                     if end.is_right_of(start)
                     else Edge(end, start))
    if shuffle:
        random.shuffle(edges)
    result = Leaf(Trapezoid(points[points_count], points[points_count + 1],
                            edges[0], edges[1]))
    for index in range(2, len(edges)):
        result = add_edge_to_graph(result, edges[index])
    return result


def add_edge_to_graph(graph: Node, edge: Edge) -> Node:
    trapezoids = find_trapezoids_intersecting_edge(graph, edge)
    p = edge.left
    q = edge.right
    left_old = None  # old trapezoid to the left.
    left_below = None  # below trapezoid to the left.
    left_above = None  # above trapezoid to the left.
    for index, old in enumerate(trapezoids):
        start_trap = index == 0
        end_trap = index == len(trapezoids) - 1
        have_left = start_trap and edge.left != old.left
        have_right = end_trap and edge.right != old.right
        left = right = None
        if start_trap and end_trap:
            if have_left:
                left = Trapezoid(old.left, p, old.below, old.above)
            below = Trapezoid(p, q, old.below, edge)
            above = Trapezoid(p, q, edge, old.above)
            if have_right:
                right = Trapezoid(q, old.right, old.below, old.above)
            if have_left:
                left.lower_left = old.lower_left
                left.upper_left = old.upper_left
                left.lower_right = below
                left.upper_right = above
            else:
                below.lower_left = old.lower_left
                above.upper_left = old.upper_left
            if have_right:
                right.lower_right = old.lower_right
                right.upper_right = old.upper_right
                below.lower_right = right
                above.upper_right = right
            else:
                below.lower_right = old.lower_right
                above.upper_right = old.upper_right
        elif start_trap:
            # Old trapezoid is the first of 2+ trapezoids
            # that the edge intersects.
            if have_left:
                left = Trapezoid(old.left, p, old.below, old.above)
            below = Trapezoid(p, old.right, old.below, edge)
            above = Trapezoid(p, old.right, edge, old.above)

            # Set pairs of trapezoid neighbours.
            if have_left:
                left.lower_left = old.lower_left
                left.upper_left = old.upper_left
                left.lower_right = below
                left.upper_right = above
            else:
                below.lower_left = old.lower_left
                above.upper_left = old.upper_left
            below.lower_right = old.lower_right
            above.upper_right = old.upper_right
        elif end_trap:
            # Old trapezoid is the last of 2+ trapezoids that the edge
            # intersects.
            if left_below.below == old.below:
                below = left_below
                below.right = q
            else:
                below = Trapezoid(old.left, q, old.below, edge)

            if left_above.above == old.above:
                above = left_above
                above.right = q
            else:
                above = Trapezoid(old.left, q, edge, old.above)

            if have_right:
                right = Trapezoid(q, old.right, old.below, old.above)

            # Set pairs of trapezoid neighbours.
            if have_right:
                right.lower_right = old.lower_right
                right.upper_right = old.upper_right
                below.lower_right = right
                above.upper_right = right
            else:
                below.lower_right = old.lower_right
                above.upper_right = old.upper_right

            # Connect to new trapezoids replacing old.
            if below != left_below:
                below.upper_left = left_below
                below.lower_left = (left_below
                                    if old.lower_left is left_old
                                    else old.lower_left)

            if above != left_above:
                above.lower_left = left_above
                above.upper_left = (left_above
                                    if old.upper_left is left_old
                                    else old.upper_left)
        else:
            # Middle trapezoid.
            # Old trapezoid is neither the first nor last of the 3+ trapezoids
            # that the edge intersects.
            if left_below.below == old.below:
                below = left_below
                below.right = old.right
            else:
                below = Trapezoid(old.left, old.right, old.below, edge)

            if left_above.above == old.above:
                above = left_above
                above.right = old.right
            else:
                above = Trapezoid(old.left, old.right, edge, old.above)

            # Connect to new trapezoids replacing prevOld.
            if below != left_below:  # below is new.
                below.upper_left = left_below
                below.lower_left = (left_below
                                    if old.lower_left is left_old
                                    else old.lower_left)

            if above != left_above:  # above is new.
                above.lower_left = left_above
                above.upper_left = (left_above
                                    if old.upper_left is left_old
                                    else old.upper_left)
            below.lower_right = old.lower_right
            above.upper_right = old.upper_right
        new_top_node = YNode(edge,
                             below.trapezoid_node
                             if below == left_below
                             else Leaf(below),
                             above.trapezoid_node
                             if above == left_above
                             else Leaf(above))
        if have_right:
            new_top_node = XNode(q, new_top_node, Leaf(right))
        if have_left:
            new_top_node = XNode(p, Leaf(left), new_top_node)
        old_node = old.trapezoid_node
        if old_node is graph:
            graph = new_top_node
        else:
            old_node.replace_with(new_top_node)
        if not end_trap:
            # Prepare for next loop.
            left_old = old
            left_above = above
            left_below = below
    return graph


def find_trapezoids_intersecting_edge(graph: Node,
                                      edge: Edge) -> List[Trapezoid]:
    trapezoid = graph.search_edge(edge)
    result = [trapezoid]
    while edge.right.is_right_of(trapezoid.right):
        orient = edge.orientation_with(trapezoid.right)
        assert orient, 'Unable to deal with point on edge.'
        if orient == -1:
            trapezoid = trapezoid.lower_right
        elif orient == 1:
            trapezoid = trapezoid.upper_right
        assert trapezoid is not None, ('Expected neighbour trapezoid, '
                                       'but none found.')
        result.append(trapezoid)
    return result
